packages/utils/src/parser/token.spec.ts

```diff
@@ -129,3 +129,305 @@ describe("ICP", () => {
+    ).toEqual(
+      TokenAmountV2.fromE8s({ token: token, amount: 100000000000000n }),
+    );
+    expect(
+      TokenAmountV2.fromString({ token: token, amount: "0.00000001" }),
+    ).toEqual(TokenAmountV2.fromE8s({ token: token, amount: 10000000000n }));
+    expect(
+      TokenAmountV2.fromString({
+        token: token,
+        amount: "0.00000000000000000001",
+      }),
+    ).toEqual(FromStringToTokenError.FractionalTooManyDecimals);
+    expect(TokenAmountV2.fromString({ token: token, amount: ".01" })).toEqual(
+      TokenAmountV2.fromE8s({
+        token: token,
+        amount: 10000000000000000n,
+      }),
+    );
+  });
+
+  it("can be initialized from a mixed string", () => {
+    expect(TokenAmountV2.fromString({ token: token, amount: "1.1" })).toEqual(
+      TokenAmountV2.fromE8s({
+        token: token,
+        amount: 1100000000000000000n,
+      }),
+    );
+    expect(TokenAmountV2.fromString({ token: token, amount: "1.1" })).toEqual(
+      TokenAmountV2.fromE8s({
+        token: token,
+        amount: 1100000000000000000n,
+      }),
+    );
+    expect(
+      TokenAmountV2.fromString({ token: token, amount: "12,345.00000001" }),
+    ).toEqual(
+      TokenAmountV2.fromE8s({
+        token: token,
+        amount: 12345000000010000000000n,
+      }),
+    );
+    expect(
+      TokenAmountV2.fromString({ token: token, amount: "12'345.00000001" }),
+    ).toEqual(
+      TokenAmountV2.fromE8s({
+        token: token,
+        amount: 12345000000010000000000n,
+      }),
+    );
+    expect(
+      TokenAmountV2.fromString({ token: token, amount: "12345.00000001" }),
+    ).toEqual(
+      TokenAmountV2.fromE8s({
+        token: token,
+        amount: 12345000000010000000000n,
+      }),
+    );
+    expect(
+      TokenAmountV2.fromString({
+        token: token,
+        amount: "100000000.000000000000000001",
+      }),
+    ).toEqual(
+      TokenAmountV2.fromE8s({
+        token: token,
+        amount: 100000000000000000000000001n,
+      }),
+    );
+    expect(
+      TokenAmountV2.fromString({
+        token: token,
+        amount: "199999999.999999999999999991",
+      }),
+    ).toEqual(
+      TokenAmountV2.fromE8s({
+        token: token,
+        amount: 199999999999999999999999991n,
+      }),
+    );
+  });
+
+  it("returns an error on invalid formats", () => {
+    expect(TokenAmountV2.fromString({ token: token, amount: "1.1.1" })).toBe(
+      FromStringToTokenError.InvalidFormat,
+    );
+    expect(TokenAmountV2.fromString({ token: token, amount: "a" })).toBe(
+      FromStringToTokenError.InvalidFormat,
+    );
+    expect(TokenAmountV2.fromString({ token: token, amount: "3.a" })).toBe(
+      FromStringToTokenError.InvalidFormat,
+    );
+    expect(
+      TokenAmountV2.fromString({ token: token, amount: "123asdf$#@~!" }),
+    ).toBe(FromStringToTokenError.InvalidFormat);
+
+    const callToNumber = () =>
+      TokenAmountV2.fromNumber({ token: token, amount: 1e-9 });
```

Comment:
I wouldn't put this under `"invalid formats"`

`1e-9` is a valid format.

Maybe add another test case `"does not support exponential notation"?

Reply:
done